File: app/src/main/java/com/zh/kotlin/singleton/java/SingleTon2.java
package com.zh.kotlin.singleton.java;

/**
 * create by zj on 2020/7/1
 * 懒加载
 */
public class SingleTon2 {
    private static SingleTon2 singleTon;

    private SingleTon2() {
    }

    public static SingleTon2 getInstance() {
        if (singleTon == null) {
            singleTon = new SingleTon2();
        }
        return singleTon;
    }
}

File: app/src/main/java/com/zh/kotlin/singleton/kotlin/SingleTon2.kt
package com.zh.kotlin.singleton.kotlin

/**
 * create by zj on 2020/7/1
 * 懒加载 线程不安全
 */
class SingleTon2 {
    companion object {
        //LazyThreadSafetyMode.NONE 没有锁
//        val instance by lazy(LazyThreadSafetyMode.NONE) {
//            SingleTon2()
//        }

        //下面是等价写法
        private var instance2: SingleTon2? = null
        fun getInstance(): SingleTon2 {
            if (instance2 == null) {
                instance2 = SingleTon2()
            }
            return instance2!!
        }


    }
}